index.js

```diff
@@ -104,21 +107,48 @@ export default e => {
       ).premultiply(app.matrixWorld).decompose(npcPlayer.position, npcPlayer.quaternion, localVector3);
       npcPlayer.updateMatrixWorld(); */
 
-      // window.npcPlayer = npcPlayer;
+      window.npcPlayer = npcPlayer; // test
 
-      if (target) {
-        const v = new THREE.Vector3().setFromMatrixPosition(target.matrixWorld)
+      if (target && npcFarawayLocalPlayer()) {
+        const isInitial = !pathFinder.destVoxel
+
+        if (isInitial || localPlayerFarawayPrevDest()) {
+          // console.log('localPlayerFarawayPrevDest')
+
+          // Don't allowNearest here, otherwise will get inaccurate result when accurate result already exists.
+          const isFound = pathFinder.getPath(npcPlayer.position, localPlayer.position, false);
+          // if (pathFinder.startVoxel) target = pathFinder.startVoxel;
+          if (isFound) target = pathFinder.waypointResult[0];
+        }
+
+        if (pathFinder.waypointResult.length > 0 && npcReachedDest()) {
+          // console.log('npcReachedDest')
+          const isFound = pathFinder.getPath(npcPlayer.position, localPlayer.position, true); // allowNearest
+          if (isFound) target = pathFinder.waypointResult[0];
+        }
+
+        if (npcReachedTarget()) {
+          // console.log('npcReachedTarget')
+          if (target._next) {
+            target = target._next;
+          }
+        }
+
+        // if (pathFinder.debugRender) console.log(target.position.x, target.position.z);
+        // const v = new THREE.Vector3().setFromMatrixPosition(target.matrixWorld)
+        const v = new THREE.Vector3().copy(target.position)
```

Comment:
This looks like a per-frame allocation. It could be removed.

Reply:
Removed. https://github.com/webaverse/npc/pull/2/commits/c95998c615a11ddcaf4cfc182eb7cb7ce014c696